### PR TITLE
Fix error in memory arbitration/reclaim in single-threaded execution

### PR DESCRIPTION
**Pull request by zhztheplayer**
Memory arbitration/reclaim works as expected in single-threaded execution in most cases except for some issues during handling task pause/resume. This patch fixes the issues. 

And enable 3 existing test cases in `SharedArbitratorTest` for single-threaded execution:

1. SharedArbitrationTest.reclaimToJoinBuilder
2. SharedArbitrationTest.reclaimToAggregation
3. SharedArbitrationTest.reclaimToOrderBy

### CODE REVIEW

File: velox/exec/Driver.cpp [hunk 1]
```diff
@@ -331,13 +331,28 @@ void Driver::pushdownFilters(int operatorIndex) {
...
+    }
+    enqueueInternal();
+    stop = runInternal(self, blockingState, result);
+    if (stop == StopReason::kPause) {
+      VELOX_DCHECK_NULL(blockingState)
+      VELOX_DCHECK_NULL(result)
+      continue;
+    }
+    break;
+  }
```
Comment (zhztheplayer) [inline velox/exec/Driver.cpp:355]: The code requires further optimization to avoid spin wait.

**Comment by zhztheplayer** [comment]
@xiaoxmeng @mbasmanova

File: velox/common/base/Exceptions.h [hunk 2]
```diff
@@ -312,6 +312,7 @@ DECLARE_CHECK_FAIL_TEMPLATES(::facebook::velox::VeloxRuntimeError);
 #define VELOX_DCHECK_LE(e1, e2, ...) VELOX_CHECK(true)
 #define VELOX_DCHECK_EQ(e1, e2, ...) VELOX_CHECK(true)
 #define VELOX_DCHECK_NE(e1, e2, ...) VELOX_CHECK(true)
+#define VELOX_DCHECK_NULL(e, ...) VELOX_CHECK(true)
```
Comment (mbasmanova) [inline velox/common/base/Exceptions.h:315]: Nice fix. Would you extract it into a separate PR?

File: velox/exec/tests/utils/SingleThreaded.cpp [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "SingleThreaded.h"
+
+namespace facebook::velox::exec::test {
+
+std::unique_ptr<folly::Executor> newSingleThreadedExecutor() {
```
Comment (mbasmanova) [inline velox/exec/tests/utils/SingleThreaded.cpp:21]: What is this for?

File: velox/common/base/Exceptions.h (hunk 2 above)
Comment (zhztheplayer) [reply velox/common/base/Exceptions.h:315]: will do

File: velox/exec/tests/utils/SingleThreaded.cpp (hunk 3 above)
Comment (mbasmanova) [reply velox/exec/tests/utils/SingleThreaded.cpp:21]: I see. I find it very confusing that a function named newSingleThreadedExecutor returns nullptr.

I suggest not to include this change in this PR.

Feel free to open a new PR to update documentation.

CC: @bikramSingh91

Comment (zhztheplayer) [reply velox/exec/tests/utils/SingleThreaded.cpp:21]: > I find it very confusing that a function named newSingleThreadedExecutor returns nullptr.

Ideally caller doesn't have to know how it's implemented but I'll remove if it feels like confusing anyway. I'll update the PR later. Thanks.

File: velox/common/base/Exceptions.h (hunk 2 above)
Comment (zhztheplayer) [reply velox/common/base/Exceptions.h:315]: Opened https://github.com/facebookincubator/velox/pull/8644

**Comment by zhztheplayer** [comment]
@mbasmanova comment addressed, thanks.

File: velox/exec/Driver.cpp [hunk 4]
```diff
@@ -331,13 +331,28 @@ void Driver::pushdownFilters(int operatorIndex) {
...
   auto self = shared_from_this();
   facebook::velox::process::ScopedThreadDebugInfo scopedInfo(
       self->driverCtx()->threadDebugInfo);
   ScopedDriverThreadContext scopedDriverThreadContext(*self->driverCtx());
   RowVectorPtr result;
-  auto stop = runInternal(self, blockingState, result);
+
+  StopReason stop;
+
+  // Spin wait when task is paused.
```
Comment (mbasmanova) [inline velox/exec/Driver.cpp:342]: Instead of spin wait, can we return a future from Task::next and have the caller come back when future completes?

Comment (zhztheplayer) [reply velox/exec/Driver.cpp:342]: That's right, I thought it should be feasible to return a future to caller somehow. But that may still require for some designs against the current common API `runInternal` and need to consider together with `BlockingState` which already have a future inside it. So I'd suggest doing the refactor in another PR so we can be flexible to design and discuss. Does that work for you? 

Also the code is currently only used when multiple independent single-thread tasks sharing the same shared arbitrator. So it just affects some use cases of the functionality.

Comment (mbasmanova) [reply velox/exec/Driver.cpp:342]: @zhztheplayer Task::next already allows for returning a Future and Driver::next supports that by producing a BlockingState. Let's figure out how to use these for this scenario. We cannot merge "spin wait" solution.

Comment (mbasmanova) [reply velox/exec/Driver.cpp:342]: @zhztheplayer Would you share a draft of a proper solution? I'm ware of landing spin-wait.

Comment (zhztheplayer) [reply velox/exec/Driver.cpp:342]: OK I understand. I am on weekend currently and will do with some initial attempts later. Thanks.
